### PR TITLE
Create `ArrayScalarBuilder` for creating single element List arrays

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

- Follow on to  https://github.com/apache/datafusion/pull/13468
- Related to https://github.com/apache/datafusion/issues/13481

## Rationale for this change

There are a bewildering number of small free functions for performing the same basic thing: converting an array into a single row list array / scalar

This has also lead to subtle bugs such as https://github.com/apache/datafusion/issues/13481

Let's encapsulate that operation into a well documented structure

## What changes are included in this PR?
1. Add `ArrayScalarBuilder`
2. Tests
3. Documentation

### CODE REVIEW

File: datafusion/common/src/utils/mod.rs [hunk 1]
```diff
@@ -321,89 +321,180 @@ pub fn longest_consecutive_prefix<T: Borrow<usize>>(
     count
 }
 
+/// Creates single element [`ListArray`], [`LargeListArray`] and
```
Comment (alamb) [inline datafusion/common/src/utils/mod.rs:324]: The idea is to pull the wrapping code into this structure so it remains consistent across list types

File: datafusion/common/src/utils/mod.rs [hunk 2]
```diff
@@ -321,89 +321,180 @@ pub fn longest_consecutive_prefix<T: Borrow<usize>>(
...
+/// // Array is [1, 2, 3]
+/// let arr = ListArray::from_iter_primitive::<Int64Type, _, _>(vec![
+///       Some(vec![Some(1), Some(2), Some(3)]),
+/// ]);
+/// // Wrap as a list array: [[1, 2, 3]]
+/// let list_arr = SingleRowArrayBuilder::new(Arc::new(arr)).build_list_array();
+/// assert_eq!(list_arr.len(), 1);
+/// ```
+#[derive(Debug, Clone)]
+pub struct SingleRowArrayBuilder {
```
Comment (jayzhan211) [inline datafusion/common/src/utils/mod.rs:344]: It seems this builder is for List, maybe name with `SingleRowListArrayBuilder` is more concise?

**Review by jayzhan211 — APPROVED** [review]

**Comment by jayzhan211** [comment]
Thanks @alamb

**Comment by alamb** [comment]
Thank you @jayzhan211

**Pull request merged**